joints/joint_collection.py

```diff
@@ -165,6 +199,39 @@ def rotate_about_axis(self, a, b, axis, theta):
             vec = self.pos[joint3] - self.pos[a]
             relative = rotate_vector(vec, axis, theta)
             self.pos[joint3] = relative + self.pos[a]
+    def get_euc_distance(self, a, b):
+        return math.sqrt( (a[0]-b[0])**2 + (a[1]-b[1])**2 + (a[2]-b[2])**2)
+     
+
+    def rotate_with_scipy(self, a, b, rotation):
+        # Save current angles
+        angles = self.get_downstream_angles(b)
+
+        # translate b to origin
+        vec_to_rotate = self.pos[b] - self.pos[a]
+        # save xyz of b before any rotation
+        b_before_rotation = self.pos[b]
+        # rotate b around origin and translate it back to have a's xyz as origin
+        self.pos[b] = rotate_scipy(vec_to_rotate, rotation) + self.pos[a]
+
+        # get the translation vector b's rotation
+        translation = self.pos[b] - b_before_rotation
+
+        # TODO !! If someone wants to optmize this, make a scipy downstream method
+        #   so you don't have to deal with this list search
+        visited = []
+        # real position of previous joint by difference of the base model      
+        for angle, _ in angles.items():
+            joint1, joint2, joint3 = angle
+            
+            # NOTE, some joints are double visited so they will not be translated again
+            if joint3 in visited:
+                continue
+            visited.append(joint3)
+            
+            # translating downsteam joints by movement of joint b
+            self.pos[joint3] = self.pos[joint3] + translation
```

Comment:
Not sure if you are monitoring this at all but just adding translation does not work and will break. You should just be able rotate all downstream joints by the same rotation.

Reply:
Hey Riley, I found that rotating all the joints downstream would break if those downstream joints, and resulting in incorrect translations and rotations. The only way I found to fix this was to rotate each joint once and adjust all downstream joints by the translation of the new - old position. If you want to try it yourself with an visualization, clone motion_predict and run this command
python .\predict_motion_test.py -t .\data\jumping_jacks_50.npy .\data\jumping_jack_50_trimmed.csv -k 1 1 -v
. Click out of the first window and the second window that comes will will show you the predicted motion rotated over time and the real points over time of the skeleton.

Reply:
I'm not totally sure why it doesn't work correctly. I am going on a cruise Saturday so I won't have a ton of time to write it until after, but we can write some tests to see if we can figure out why it doesn't work because this method won't preserve downstream angles. See this drawing as to why. See how the red angle changes after just translating the blue point instead of also rotating it. 
[Note Jan 31, 2024.pdf](https://github.com/JMU-Wearable-Computing/joints/files/11512505/Note.Jan.31.2024.pdf)

Reply:
Hey, I am making some tests right now, I'll just upload the file and the data used for them here by Friday night, and I'll add what I think you originally envisioned as the scipy rotation methodology. I'll be using the forward_kinematics.py file in motion_predict to rotate the sets of rotations supplied.

Reply:
Haven't finished the tests, but here are three images showing the examples of different ways to rotate it. 
![base](https://github.com/JMU-Wearable-Computing/joints/assets/92378119/3dd57b4e-63d9-4a2b-b6a9-dcfe129ada64)